### DISCUSSION TITLE
[Feature Request] Type definition for generators, like "image", "video", "pdf", "audio".

### DISCUSSION THREAD

**Discussion by bcalik** [spatie/laravel-medialibrary#1953] [Ideas] [answered]
Since this is a "media" library, it's better support all kind of media, not only images.

We have ImageGenerators, then we can have PdfGenerators, VideoGenerators, AudioGenerators etc.
Users should be able to define its own types and generators.

Then I would like to define multiple conversions and specify which generator to use.  
An example would be like this;

```php
// To generate thumbnail conversions.
$this->addMediaConversion('thumb')
     ->usingGenerator('image') // which will be default
     ->width(200)
     ->height(200);

// To generate pdf conversions.
$this->addMediaConversion('pdf')
     ->usingGenerator('pdf');
```

So the uploaded file will have both thumbnail and pdf conversions ready to use.

**Comment by freekmurze** [comment]
You can already define generators which can determine themselves if they should be used.

https://docs.spatie.be/laravel-medialibrary/v8/converting-other-file-types/creating-a-custom-image-generator/

**Comment by bcalik** [comment, marked as answer]
@freekmurze I think you get me wrong.  
I already know I can define custom generators, but they are **image** generators.  
Image generators are for to generate **images**. But I don't want to generate images.

I want to generate anything else.

For example I would upload a MP4 file and generate an MP3 from it.
That would be called AudioGenerator, with type "audio", not "image".

That's why I said "generator **TYPE** definition" at the first place.

**Reply by freekmurze** [reply]
Right, for now this kind of conversions are out of scope for the media library. Three reasons:
- the package is already very feature rich and I don't want to tack on this extra functionality
- not many people have asked for this
- I don't need this myself


Please convert things to a format you want before adding them to the media library.

**Comment by AlejandroAkbal** [comment]
Hey @bcalik

I'm having the same issue and would like to know with which solution you have ended up

What I'm doing right now is pre-processing videos before uploading them
But feels shitty since the most elegant solution would be to create a generator/conversion that does this and integrates with Laravel media library
